Introduced parcel based frontend build process

naltatis opened the pull request:
Here's the first shot to move the frontend setup to the next level. I've tried to change as little files in this PR. Only updated what's necessary to introduce a lightweight build process. This is the prerequisite for topics like single-file components, component tests, linting and code formatting in the frontend.

## changes
- added [parcel](https://parceljs.org) as a minimal frontend bundler
- removed vendor code (vue, vue-router,  axios) from repo and added it as NPM dependencies
- fixed some invalid HTML and Vue warnings
- adapted Go's asset generation process to match new output directory

## improvements
- hot module replacement and proper error messages when running frontend in dev-mode (`npm run dev`)
- content based file hashing (for proper HTTP cache invalidation)
- optimised and minified asset in production build

## known issues
- jQuery, bootstrap.js and popper are still in-repo dependencies. I'd recommend to replace them with a proper vue-bootstrap implementation to reduce overall bundle size and avoid conflicts between jQuery's imperative and Vue's declarative approach to DOM manipulation
- site.webmanifest produces incorrect paths to the referenced icons. see https://github.com/parcel-bundler/parcel/issues/3588. I'll address this in a separate PR.

andig commented on .gitignore:
```diff
@@ -1,11 +1,12 @@
 __debug_bin
 .vscode
-*.json
```
Warum? Verhindert dass aus versehen Prod Configs eingecheckt werden.

naltatis replied:
Ich kann den auch spezieller machen. Konkret wurde die package[-lock].json ignoriert und gerade in der FE-Welt ist das auch nicht unüblich .json Files bspw. für i18n zu haben.

Welche Prod Configs sind das denn? Hatte im Projekt gesucht und konnte kein anderes JSON finden.

andig replied:
Deshalb ja ;) Ich habe oft lokal tokens uns ähnliches liegen das nicht ins repo darf. Du kannst m.E. *.json ausschliessen und package*.json wieder erlauben?

naltatis replied:
✅ done

andig commented on assets/index.html:
```diff
@@ -80,7 +73,7 @@
   </nav>
 
-  <version installed="<<.Version>>" id="version-bar"></version>
+  <version installed="[[.Version]]" id="version-bar"></version>
```
Super Idee mit den [[ ]]. Bin ich irgendwie nicht drauf gekommen :)

naltatis replied:
Mal ne andere Frage. Was macht das `<version>` Tag? Konnte dafür keine Vue-Komponente finden.

andig replied:
Gute Frage. Soll egtl. das Ergebnis des git version checks anzeigen. Muss mal probieren ob ich das kaputt gemacht habe :O

andig replied:
Hier isse: https://github.com/andig/evcc/blob/master/assets/js/app.js#L234

andig commented on assets/index.html:
```diff
@@ -519,8 +513,7 @@ <h2 class="value">
   </div>
 </footer>
-
-<script src="js/app.js?<<.Tag>>"></script>
```
.Tag brauchen wir damit iOS Web Apps sich refreshen können. Egtl. auch analog fürs Icon, ggf. mittels Hash im Backend?

naltatis replied:
Das explizite Tagging ist mit Parcel nicht mehr notwendig da er automatisch alle Assets mit Content-Hashes versieht.

naltatis replied:
In der generierten `dist/assets/index.html` siehst du dann auch die angepassten Referenzen.
`<script src="/dist/app.f170d70e.js"></script>`

andig replied:
Ahhh, nice 👍🏻. Es könnte aber sein, dass embed das nicht versteht, ich glaube es „sieht“ neue Dateien nicht. Muss ich probieren. Falls das der Fall sein sollte müssten wir auf Go 1.16. warten.

Ist der Build deterministisch, also werden immer die gleichen Referenzen bei gleichem Content erzeugt?

naltatis replied:
Ja, eigentlich sollte der Assets Build deterministisch sein. Lokal auf meiner Maschine war das auch immer so. Im vorherigen Travis Lauf scheint sich aber die assets.go geändert zu haben. Daher hab ich sie auch ins ignore gepackt.
Bin mir nicht sicher welche Node/NPM Version in Travis gezogen wird oder das OS noch eine Rolle spielt.

Was meinst du mit dem `clean` Check und dass "embed" das nicht versteht?

andig replied:
> Im vorherigen Travis Lauf scheint sich aber die assets.go geändert zu haben.

Die Generierung der assets.go ist deterministisch- dann liegt das Problem vorher. 

> Was meinst du mit dem clean Check und dass "embed" das nicht versteht?

Ich meine https://github.com/andig/evcc/blob/master/.travis.yml#L31. Da wird gecheckt ob notwendige generierte Artefakte sauber im git eingecheckt sind, d.h. also ob beim assets generieren wieder das gleiche Ergebnis raus kommt.

naltatis replied:
Ich hab jetzt noch mal sichergestellt, dass gehashte Assets aus vorherigen Builds auch abgeräumt werden. Wenn Travis den Workspace nicht abräumt kann es gut sein, das das auch der Grund war warum die assets.go beim zweiten Durchlauf modifiziert wurde.

andig commented on assets/js/app.js:
```diff
@@ -129,7 +135,7 @@ window.setInterval(function() {
 //
 
-const main = Vue.component('main', {
+const main = Vue.component('Main', {
```
Best practice?

naltatis replied:
Ja, gernell kenn ich das so, dass man Vue Komponenten groß schreibt. Hier gabs konkret das Problem, dass es in HTML 5 ein `main` Element gibt und Vue über den Konflikt meckert.

andig replied:
Dann sollten wir das konsequent für alle Komponenten machen?

naltatis replied:
Ja, sollten wir. Wollte ich aber nicht in diesem PR machen und hatte kein schöneres Wort als "main" parat. Ich würde ja sowieso gerne perspektivisch auf Single File Componentes refactorn. Das wäre dann sicher ein guter Zeitpunkt für die Namensänderung.

andig commented:
Nice- jetzt muss ich das ausprobieren :O. Wie würde ich das nutzen um weiter "live" entwickeln zu können? `npm run dev` im Rootfolder?

naltatis commented:
> Nice- jetzt muss ich das ausprobieren :O. Wie würde ich das nutzen um weiter "live" entwickeln zu können? `npm run dev` im Rootfolder?

Ja, du kannst `npm run build  --prefix assets` aus dem Root Folder machen. Alternativ auch ohne `--prefix` direkt im `assets` Verzeichnis. Vorher natürlich einmal ein `npm install`. Kannst dir auch `make build-assets` Task anschauen.

andig commented:
Du hast die Assets jetzt rausgenommen. Aktuell sind die immer drin damit ein einfaches `make build` geht und immer sicher auf einen sauberen Stand aufsetzt?

naltatis commented:
> Du hast die Assets jetzt rausgenommen. Aktuell sind die immer drin damit ein einfaches `make build` geht und immer sicher auf einen sauberen Stand aufsetzt?

Ja, verstehe ich. Nach der Änderung muss man als Entwickler Node installiert und `build-assets` ausgeführt haben damit man auch UI sieht.

Die `assets.go` eingecheckt lassen halte ich für schwierig weil ich davon ausgehe, dass wir hier zu regelmäßigen Mergekonflikten kommen. Das müsste aber doch auch bis jetzt schon mal ein Problem gewesen sein, oder?

andig commented:
> Die assets.go eingecheckt lassen halte ich für schwierig weil ich davon ausgehe, dass wir hier zu regelmäßigen Mergekonflikten kommen. Das müsste aber doch auch bis jetzt schon mal ein Problem gewesen sein, oder?

Solange es linear abläuft kein Problem. Bei Rebase und Konflikt lösche ich die dann einfach, muss aber gemacht werden. 

Sie eingecheckt zu haben erleichtert den lokalen Build "mal eben". @TheRealKerni kannst Du Dich erinnern warum wir das noch haben wollten?

andig commented on assets/package.json:
```diff
@@ -0,0 +1,26 @@
+{
```
Wäre evtl möglich das ins root zu ziehen damit alles aus einem Ordner raus machbar ist?

naltatis replied:
Ja klar. Kann ich umstellen.

naltatis replied:
Hab das gerade umgebaut. `package*json` liegt jetzt im Root-Verzeichnis. Keine Notwendigkeit mehr für Prefixes. Hatte gerade auch noch das Problem, dass die generierte assets.go im Dev-Mode nicht mit neu erzeugten Dateien klargekommen ist. Daher habe ich jetzt ne explizite Fallunterscheidung in die `http.go` eingebaut. Schau dir mal bitte an, ob man das so machen kann.

andig commented:
Mhhm, das ist auch doof:

```
❯ gco master 
Alias tip: gcm
error: Your local changes to the following files would be overwritten by checkout:
        assets/package-lock.json
Please commit your changes or stash them before you switch branches.
Aborting
```

naltatis commented:
> Mhhm, das ist auch doof:
> 
> ```
> ❯ gco master 
> Alias tip: gcm
> error: Your local changes to the following files would be overwritten by checkout:
>         assets/package-lock.json
> Please commit your changes or stash them before you switch branches.
> Aborting
> ```

Sollte nicht so sein. Was ist denn die Änderung die er anmeckert? Welche Node/NPM Version hast du?

andig commented:
```
❯ npm --version
7.0.3

~/htdocs/evcc naltatis/frontend_build*
❯ npm i

added 758 packages, and audited 758 packages in 13s

2 high severity vulnerabilities

To address all issues (including breaking changes), run:
  npm audit fix --force

Run `npm audit` for details.

~/htdocs/evcc naltatis/frontend_build* 14s
❯ npm run dev  

> dev
> rimraf dist; parcel watch assets/index.html --public-url /dist/

✨  Built in 10.81s.
```

gibt 200k diff im package-lock?

andig commented:
Oder wir lassen das file weg?

naltatis commented:
Best Practice ist eigentlich die package-lock.json mit einzuchecken um Reproduzierbarkeit zu haben.

Habs jetzt selbst mal probiert. Scheinbar führt NPM mit v7 (ich hatte noch v6) ein neues Package-Lock Format ein. Daher der so massive Diff. Ich generier den mal im Branch neu.

```
 {
   "name": "evcc",
+  "lockfileVersion": 2,
   "requires": true,
-  "lockfileVersion": 1,
+  "packages": {
```

TheRealKerni commented:
> Sie eingecheckt zu haben erleichtert den lokalen Build "mal eben". @TheRealKerni kannst Du Dich erinnern warum wir das noch haben wollten?

Ne leider nicht. Falls aber damit ein Problem auftritt, wird es wieder passieren und dann wissen wir es :)

naltatis commented:
I switched to node 14 (and thereby NPM 6) to avoid the new package-lock.json format for now. I was unable to convince Travis CI to use node 15. I've adapted the readme accordingly and added a version constrain to the package.json. So you should downgrade to node@14 to avoid future conflicts.

Irgendwie bin ich immer im englisch-Modus wenn ich auf Github bin. Aber ich übersetz das jetzt nicht wieder zurück. You get the point ;)

andig committed:
Fix site api error

andig committed:
Add REST keyword

andig committed:
Improve error handling and provide better error message

andig committed:
Add evcc.dist.yaml to release packages

andig commented on assets/index.html:
```diff
@@ -29,13 +29,6 @@
   <script src="js/popper.min.js"></script>
   <script src="js/bootstrap.bundle.min.js"></script>
```
Bootstrap und Popper analog vue und axios?

naltatis replied:
Ja, könnte man auch jetzt schon zu einer Dependency machen. Hatte das ja in "know issues" auch erwähnt jQuery würd ich aber erstmal als Script lassen und dann später durch https://bootstrap-vue.org ersetzen.
Ich update den PR heute Abend.

andig replied:
Îm Prinzip könnten wir‘s ja auch lassen. Bootstrap 5 steht schon vor der Tür, ggf. Lohnt es sich auch nicht hier zu investieren. Mit bootstrap-vue habe ich mich noch nicht auseinander gesetzt, es schien mir aber durchaus komplex und eben deutlich weniger dokumentiert als Bopotstrap selbst. Würde uns das wirklich helfen?

naltatis replied:
Ich habe bootstrap-vue auch noch nicht selbst verwendet. Der große Vorteil ist halt, dass man dann klar beim deklarativem Paradigma von Vue bleibt und es nicht mit dem imperativen von jQuery mischt. Aber ich habe den bisherigen Frontend Code noch nicht genau genug gelesen/verstanden. Daher würde ich diese Entscheidung auch gerne erstmal verschieben.

naltatis commented:
So, hab jetzt auch die anderen CSS/JS/Font Vendor Libraries auf NPM umgestellt. Nun liegt nur noch selbstgeschriebener Code im `assets` Verzeichnis.

andig committed:
Fix configuration for updated Renault error handling

naltatis commented:
@andig Ist aus deiner Sicht noch was offen?

andig commented:
Nein- wenn Du mit npm zufrieden bist würde ichs mal ausprobieren?

andig committed:
Improve user interface when minSoC is active (#393)

andig committed:
Auto-detect connected vehicle (#397)

andig commented:
Mist, ich hätte nicht mergen dürfen. Würdest Du nochmal einen Rebase machen? Sorry für den Konflikt, das war ohne Not :(

andig commented on Makefile:
```diff
@@ -25,6 +25,10 @@ test:
 	@echo "Running testsuite"
 	go test ./...
 
+build-assets:
```
Willst Du evtl hier nochmal was am naming machen? Wir haben jetzt build, build-assets und assets. Wie wärs hier einfach mit ui?

naltatis replied:
Ja find ich gut. Ich benenne den Task in `ui` um und löse die Konflikte.

andig commented on assets/index.html:
```diff
@@ -21,46 +21,9 @@
-    }
-  </style>
+  <link href="../node_modules/bootstrap/dist/css/bootstrap.css" rel="stylesheet">
+  <link href="../node_modules/@fortawesome/fontawesome-free/css/all.css" rel="stylesheet">
```
das kann eigentlich nicht stimmen- `fortawesome`?

naltatis replied:
Hat mich auch irritiert, aber das scheint die offizielle NPM Organisation zu sein. Siehe https://fontawesome.com/how-to-use/on-the-web/setup/using-package-managers

andig committed:
Fix charger wrongly enabled on startup when vehicle connected and minsoc defined (#402)

andig committed:
Maintain mode=off on disconnect (#403)

andig committed:
Improve formatting

andig committed:
Fix minSoc remainig charge displayed when soc unknown

andig committed:
Update core logger name to site to match configuration

andig commented:
Also:

- npm i würde ich nach make install verschieben
- npm zerschiesst bei mir weiter die package log
- unter config sehe ich --> (komisch?), im Code ists egtl nur ein Kommentar der für mich regulär aussieht -> Fehler im Builder/Packager?
- wenn ich die Zeilen rauslösche während npm run dev läuft, zerschiesst es mir nach Reload alles:
<img width="333" alt="Screenshot 2020-11-06 at 18 18 33" src="https://user-images.githubusercontent.com/184815/98395504-a4dda000-205c-11eb-86ac-475dda762830.png">
<img width="842" alt="Screenshot 2020-11-06 at 18 21 43" src="https://user-images.githubusercontent.com/184815/98395708-eff7b300-205c-11eb-9661-31f2f00a43cf.png">

- prinzipiell wäre es im dev modus (npm run dev) schön wenn das UI auch alleine refreshen würde?

Noch ist der Flow nicht ganz flüssig?

naltatis commented:
> npm i würde ich nach make install verschieben
Bin nicht sehr `Makefile` versiert. Solange `npm install` immer vor den anderen `npm` commands aufgerufen wird passt das ja. -> ist geändert

> npm zerschiesst bei mir weiter die package log
Hast du dein node auf v14 gedowngraded? Das hatte ich ja in der Readme / package.json entsprechend angepasst.

> prinzipiell wäre es im dev modus (npm run dev) schön wenn das UI auch alleine refreshen würde?
Die UI refresht sich bei mir im dev-Mode. Bekommst du Fehler in der Browser-Konsole?

> unter config sehe ich --> (komisch?), im Code ists egtl nur ein Kommentar der für mich regulär aussieht -> Fehler im Builder/Packager?
Hier scheint der HTML Minifzierer ein Problem mit HTML Kommentaren in Script Tags (ist auch ne komische Kombi) zu haben.  Habe ihn jetzt so konfiguriert, dass er die Kommentare einfach drin lässt. Mit Single-File Components ist das später sowieso kein Problem mehr :)

naltatis committed:
introduced frontend build process via parcel; managing FE dependencies via NPM; fixed some HTML issues and Vue warnings

naltatis commented:
🎈

andig commented:
Es ist aber immer noch zickig:

    🚨  Cannot read property 'type' of undefined
        at Bundler.createBundleTree (/Users/andig/htdocs/evcc/node_modules/parcel-bundler/src/Bundler.js:654:54)
        at Bundler.createBundleTree (/Users/andig/htdocs/evcc/node_modules/parcel-bundler/src/Bundler.js:721:12)
        at Bundler.bundle (/Users/andig/htdocs/evcc/node_modules/parcel-bundler/src/Bundler.js:298:14)
        at processTicksAndRejections (internal/process/task_queues.js:93:5)
    make: *** [install] Interrupt: 2

ich probier mal ein wenig...

andig commented:
Ah. man darf `npm run dev` nicht parallel zu make offen haben. Schöne neue Welt ;)

naltatis commented:
> Ah. man darf `npm run dev` nicht parallel zu make offen haben. Schöne neue Welt ;)
> 
> 

Heißt da lief npm run build und npm run dev parallel?

andig commented:
...weil ich das run dev noch im Hintergrund hatte um rumzubasteln... Räume jetzt erstmal die Dockerfiles auf damit der Build auch da wieder läuft- das konntest Du im PR nicht sehen...
